xmpp/xmpp.go

```diff
@@ -655,14 +656,48 @@ func Dial(address, user, domain, password string, config *Config) (c *Conn, err
 			return nil, errors.New("xmpp: session establishment failed")
 		}
 	}
+
+	go c.ping(domain)
 	return c, nil
 }
 
-func makeInOut(conn io.ReadWriter, config *Config) (in *xml.Decoder, out io.Writer) {
+const pingTime = 1 * time.Minute
+const ioTimeout = 3 * time.Minute
+
+// Send an XMPP ping periodically to verify the connection is alive.
+func (c *Conn) ping(domain string) {
+	for {
+		time.Sleep(pingTime)
+		fmt.Fprintf(c.out,
```

Comment:
This will cause concurrent writes to c.out, but the other writes in the file aren't atomic. Either this code needs a lot of locking (ugh) or else this needs to occur at a higher level, in the user-interaction code.